meegkit/ress.py

```diff
@@ -29,92 +26,236 @@ def RESS(X, sfreq: int, peak_freq: float, neig_freq: float = 1,
+            self._gaussfilt(
+                X, self.sfreq, self.peak_freq, fwhm=self.peak_width, n_harm=1
+            )
+        )
+
+        # add 1% regularization to avoid numerical precision problems in the GED
+        c0 = (c01 + c02) / 2
+        c0 = c0 * (1 - self.gamma) + self.gamma * np.trace(c0) / len(c0) * np.eye(
+            len(c0)
+        )
+
+        # Perform generalized eigendecomposition
+        # It solve a Generalized Eigenvalue Problem:
+        # find a vector `to_ress` that maximize the ratio
+        # c0^{-1} c1 (max c1 and min c0)
+        d, to_ress = linalg.eigh(c1, c0)
+        # Keep only the real part as c1 and c0
+        # are symetric (then PSD) the imaginary
+        # part is only a numerical error
+        d = d.real
+        to_ress = to_ress.real
+
+        # Sort eigenvectors by decreasing eigenvalues
+        # We are looking for the eigenvectors associated
+        # with the larger eigenvalue
+        idx = np.argsort(d)[::-1]
+        d = d[idx]
+        to_ress = to_ress[:, idx]
+
+        # Normalize components (yields mixing matrix)
+        to_ress /= np.sqrt(np.sum(to_ress, axis=0) ** 2)
+        self.to_ress = to_ress[:, np.arange(self.n_keep)]
+
+        # Compute unmixing matrix
+        A = np.matmul(c1, to_ress)
+        B = np.matmul(to_ress.T, np.matmul(c1, to_ress))
+        try:
+            # Note: we must use overwrite_a=False in order to be able to
+            # use the fall-back solution below in case a LinAlgError is raised
+            from_ress = linalg.solve(B.T, A.T, overwrite_a=False)
+        except linalg.LinAlgError:
+            # A or B is not full rank so exact solution is not tractable.
+            #  Using least-squares solution instead.
+            from_ress = linalg.lstsq(B.T, A.T, lapack_driver="gelsy")[0]
+        self.from_ress = from_ress[: self.n_keep, :]
+
+        return self
+
+    def transform(self, X):
+        """Project data using the learned RESS filter"""
+
+        if X.ndim == 2:
+            out = np.matmul(X, self.to_ress)
+        else:
+            out = np.zeros((X.shape[0], self.n_keep, X.shape[2]))
+            for idx_trial in range(len(X)):
+                out[idx_trial] = np.matmul(X[idx_trial].T, self.to_ress).T
+
         return out
+
+    def fit_transform(self, X, y=None):
+
+        self.fit(X)
+
+        return self.transform(X)
+
+    def inverse_transform(self, X):
+        """Backproject the RESS filtered data to the sensors space"""
+
+        return np.matmul(X, self.from_ress)
+
+    def _avg_cov(self, X):
+        """Compute average covariance matrix along trials"""
+        X_transpose = X.transpose([0, 2, 1])
+        s = list(X_transpose.shape)
+        combined = functools.reduce(lambda x, y: x * y, s[0: 0 + 1 + 1])
+        X_combined = np.reshape(X_transpose, s[:0] + [combined] + s[0 + 1 + 1:])
+        cov = EmpiricalCovariance().fit(X_combined)
+        return cov.covariance_
+
+    def _gaussfilt(
```

Comment:
I have other projects that use the gaussfilt function (not related to RESS at all). Can you not keep this in a separate function?

Reply:
BTW I thing there is something strange like a redundant copy-paste in `gaussifilt` function (line 338) but maybe this is correct ?